File: ios/FlowerLines/Layout/LayoutCalculator.swift
import UIKit

struct Layout {
    let cellSize: CGFloat
    let scale: CGFloat
    let flowerRadius: CGFloat
    let sidebarFlowerRadius: CGFloat
    let headerHeight: CGFloat
    let sidebarWidth: CGFloat
    let gridWidth: CGFloat
    let totalWidth: CGFloat
    let totalHeight: CGFloat
    let offsetX: CGFloat   // horizontal centering offset
    let offsetY: CGFloat   // vertical centering offset

    // Scaled font sizes
    var scoreFontSize: CGFloat    { 19 * scale }
    var bestFontSize: CGFloat     { 15 * scale }
    var buttonFontSize: CGFloat   { 14 * scale }
    var sidebarTitleFont: CGFloat { 13 * scale }
    var gameOverTitleFont: CGFloat{ 34 * scale }
    var gameOverBodyFont: CGFloat { 22 * scale }
    var gameOverBtnFont: CGFloat  { 18 * scale }

    // Button rects (in local layout coords, before offsetting)
    var newGameBtn: CGRect {
        CGRect(x: 340*scale, y: 12*scale, width: 150*scale, height: 36*scale)
    }
    var volumeBtn: CGRect {
        CGRect(x: 290*scale, y: 12*scale, width: 36*scale, height: 36*scale)
    }
    var gobAdBtn: CGRect {
        CGRect(x: 142*scale, y: 340*scale, width: 220*scale, height: 44*scale)
    }
    var gobBtn: CGRect {
        CGRect(x: 142*scale, y: 394*scale, width: 220*scale, height: 38*scale)
    }

    func cellCX(_ col: Int) -> CGFloat { CGFloat(col) * cellSize + cellSize/2 }
    func cellCY(_ row: Int) -> CGFloat { headerHeight + CGFloat(row) * cellSize + cellSize/2 }
    func cellRect(_ row: Int, _ col: Int) -> CGRect {
        CGRect(x: CGFloat(col)*cellSize, y: headerHeight + CGFloat(row)*cellSize,
               width: cellSize, height: cellSize)
    }
}

func calculateLayout(bounds: CGRect, safeAreaInsets: UIEdgeInsets) -> Layout {
    // Grid fills 100% of screen width; no sidebar
    let screenW  = bounds.width
    let usableH  = bounds.height - safeAreaInsets.top - safeAreaInsets.bottom

    let cellSize = screenW / CGFloat(GRID_SIZE)
    let scale    = cellSize / 56.0

    let sidebarWidth: CGFloat = 0
    let headerHeight  = 60.0  * scale
    let gridWidth     = cellSize * CGFloat(GRID_SIZE)
    let totalWidth    = gridWidth + sidebarWidth
    let totalHeight   = headerHeight + cellSize * CGFloat(GRID_SIZE)

    let offsetX: CGFloat = 0
    let offsetY = safeAreaInsets.top + max((usableH - totalHeight) / 2, 0)

    return Layout(
        cellSize: cellSize, scale: scale,
        flowerRadius: 20 * scale, sidebarFlowerRadius: 22 * scale,
        headerHeight: headerHeight, sidebarWidth: sidebarWidth,
        gridWidth: gridWidth, totalWidth: totalWidth, totalHeight: totalHeight,
        offsetX: offsetX,
        offsetY: offsetY
    )
}

File: android/app/src/main/java/com/flowerlines/layout/LayoutCalculator.kt
package com.flowerlines.layout

import android.graphics.RectF
import com.flowerlines.game.GRID_SIZE

data class Layout(
    val cellSize: Float,
    val scale: Float,
    val flowerRadius: Float,
    val sidebarFlowerRadius: Float,
    val headerHeight: Float,
    val sidebarWidth: Float,
    val gridWidth: Float,
    val totalWidth: Float,
    val totalHeight: Float,
    val offsetX: Float,
    val offsetY: Float
) {
    val scoreFontSize:     Float get() = 19f * scale
    val bestFontSize:      Float get() = 15f * scale
    val buttonFontSize:    Float get() = 14f * scale
    val sidebarTitleFont:  Float get() = 13f * scale
    val gameOverTitleFont: Float get() = 34f * scale
    val gameOverBodyFont:  Float get() = 22f * scale
    val gameOverBtnFont:   Float get() = 18f * scale

    val newGameBtn: RectF  get() = RectF(340f*scale, 12f*scale, (340f+150f)*scale, (12f+36f)*scale)
    val volumeBtn:  RectF  get() = RectF(290f*scale, 12f*scale, (290f+36f)*scale,  (12f+36f)*scale)
    val gobAdBtn:   RectF  get() = RectF(142f*scale, 340f*scale,(142f+220f)*scale, (340f+44f)*scale)
    val gobBtn:     RectF  get() = RectF(142f*scale, 394f*scale,(142f+220f)*scale, (394f+38f)*scale)

    fun cellCX(col: Int) = col.toFloat() * cellSize + cellSize / 2f
    fun cellCY(row: Int) = headerHeight + row.toFloat() * cellSize + cellSize / 2f
}

fun calculateLayout(
    viewWidth: Int, viewHeight: Int,
    insetLeft: Int = 0, insetTop: Int = 0, insetRight: Int = 0, insetBottom: Int = 0
): Layout {
    val usableW = viewWidth  - insetLeft - insetRight
    val usableH = viewHeight - insetTop  - insetBottom

    val cellFromWidth  = usableW  / (GRID_SIZE + 130f / 56f)
    val cellFromHeight = usableH / (GRID_SIZE + 60f  / 56f)
    val cellSize = minOf(cellFromWidth, cellFromHeight)
    val scale    = cellSize / 56f

    val sidebarWidth = 130f * scale
    val headerHeight = 60f  * scale
    val gridWidth    = cellSize * GRID_SIZE
    val totalWidth   = gridWidth + sidebarWidth
    val totalHeight  = headerHeight + cellSize * GRID_SIZE

    val offsetX = insetLeft  + (usableW - totalWidth)  / 2f
    val offsetY = insetTop   + (usableH - totalHeight) / 2f

    return Layout(
        cellSize = cellSize, scale = scale,
        flowerRadius = 20f * scale, sidebarFlowerRadius = 22f * scale,
        headerHeight = headerHeight, sidebarWidth = sidebarWidth,
        gridWidth = gridWidth, totalWidth = totalWidth, totalHeight = totalHeight,
        offsetX = maxOf(offsetX, insetLeft.toFloat()),
        offsetY = maxOf(offsetY, insetTop.toFloat())
    )
}
